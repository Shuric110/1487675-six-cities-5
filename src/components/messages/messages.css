.app-messages {
  display: block;
  list-style: none;
  padding: 0;
  position: absolute;
  top: 25px;
  right: 50px;
  width: 350px;
  z-index: 1000;
}

.app-messages__item {
  padding: 20px;
  border-radius: 15px;
  margin: 0 0 25px;
  opacity: 1;
}

.app-messages__item--error {
  background-color: rgba(255, 100, 100, 0.4);
  color: #cc0000;
}

.app-messages__item--fadeout {
  transition: opacity 1s ease-in-out;
  opacity: 0;
}
