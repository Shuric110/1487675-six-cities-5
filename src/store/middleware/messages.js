import {ActionType} from "../action";
import {extend} from "../../util";
import {MESSAGE_SHOW_TIME, MESSAGE_FADEOUT_TIME} from "../../const";
import {ActionCreator} from "../action";

const idSequence = {
  _currentId: 1,

  getNextId() {
    return this._currentId++;
  }
};

export const messages = ({dispatch}) => (next) => (action) => {
  if (action.type === ActionType.SHOW_MESSAGE) {
    const id = idSequence.getNextId();
    action = extend(action, {payload: extend(action.payload, {
      id,
      isFadingOut: false,
    })});

    setTimeout(() => {
      dispatch(ActionCreator.fadeoutMessage(id));
    }, MESSAGE_SHOW_TIME);

    setTimeout(() => {
      dispatch(ActionCreator.removeMessage(id));
    }, MESSAGE_SHOW_TIME + MESSAGE_FADEOUT_TIME);
  }

  return next(action);
};
