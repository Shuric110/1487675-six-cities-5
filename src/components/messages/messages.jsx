import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import "./messages.css";

const Messages = (props) => {
  const {messages} = props;

  return (
    <ul className="app-messages">
      {messages.map(({id, text, isFadingOut}) =>
        <li key={id} className={`app-messages__item app-messages__item--error ${isFadingOut ? `app-messages__item--fadeout` : ``}`}>
          {text}
        </li>
      )}
    </ul>
  );
};

Messages.propTypes = {
  messages: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    isFadingOut: PropTypes.bool.isRequired,
  })).isRequired
};


const mapStateToProps = (state) => ({
  messages: state.APP.messages,
});

export {Messages};
export default connect(mapStateToProps)(Messages);
