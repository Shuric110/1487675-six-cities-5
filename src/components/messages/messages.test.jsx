import React from "react";
import renderer from "react-test-renderer";
import {Messages} from "./messages";

const messages = [
  {
    id: 1,
    text: `Message 1`,
    isFadingOut: true,
  },
  {
    id: 2,
    text: `Message 2`,
    isFadingOut: false,
  },
  {
    id: 3,
    text: `Message 3`,
    isFadingOut: false,
  },
];

describe(`Render Messages`, () => {
  it(`Messages renders correctly`, () => {
    const tree = renderer
      .create(
          <Messages
            messages={messages}
          />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
