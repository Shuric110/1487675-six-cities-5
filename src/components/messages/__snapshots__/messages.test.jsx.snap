// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Messages Messages renders correctly 1`] = `
<ul
  className="app-messages"
>
  <li
    className="app-messages__item app-messages__item--error app-messages__item--fadeout"
  >
    Message 1
  </li>
  <li
    className="app-messages__item app-messages__item--error "
  >
    Message 2
  </li>
  <li
    className="app-messages__item app-messages__item--error "
  >
    Message 3
  </li>
</ul>
`;
